src/servers/src/otlp/logs.rs

```diff
@@ -355,35 +358,343 @@ fn build_identity_row(
+            ))),
+            JsonbNumber::UInt64(u) => Ok(Some((
+                GreptimeValue {
+                    value_data: Some(ValueData::U64Value(u)),
+                },
+                ColumnDataType::Uint64,
+                SemanticType::Tag,
+                None,
+            ))),
+        },
+        JsonbValue::Bool(b) => Ok(Some((
+            GreptimeValue {
+                value_data: Some(ValueData::BoolValue(b)),
+            },
+            ColumnDataType::Boolean,
+            SemanticType::Field,
+            None,
+        ))),
+        JsonbValue::Array(_) | JsonbValue::Object(_) => OpenTelemetryNestValueIsComplicatedSnafu {
+            key: column_name.as_str(),
+        }
+        .fail(),
+        JsonbValue::Null => Ok(None),
+    };
+    column_info.map(|c| {
+        c.map(|(value, column_type, semantic_type, datatype_extension)| {
+            (
+                ColumnSchema {
+                    column_name,
+                    datatype: column_type as i32,
+                    semantic_type: semantic_type as i32,
+                    datatype_extension,
+                    options: None,
+                },
+                value,
+            )
+        })
+    })
+}
+
+#[derive(Debug)]
+enum OpenTelemetryLogRecordAttrType {
+    Resource,
+    Scope,
+    Log,
+}
+
+fn merge_schema<'a>(
+    extra_resource_schema: &'a SchemaInfo,
+    extra_scope_schema: &'a SchemaInfo,
+    extra_log_schema: &'a SchemaInfo,
+) -> BTreeMap<&'a String, (OpenTelemetryLogRecordAttrType, usize, &'a ColumnSchema)> {
+    let mut schemas = BTreeMap::new();
+    for (resource_schema_key, resource_schema_index) in extra_resource_schema.index.iter() {
+        if let Some(column_schema) = extra_resource_schema.schema.get(*resource_schema_index) {
+            schemas.insert(
+                resource_schema_key,
+                (
+                    OpenTelemetryLogRecordAttrType::Resource,
+                    *resource_schema_index,
+                    column_schema,
+                ),
+            );
+        }
+    }
+    for (scope_schema_key, scope_schema_index) in extra_scope_schema.index.iter() {
+        if let Some(column_schema) = extra_scope_schema.schema.get(*scope_schema_index) {
+            schemas.insert(
+                scope_schema_key,
+                (
+                    OpenTelemetryLogRecordAttrType::Scope,
+                    *scope_schema_index,
+                    column_schema,
+                ),
+            );
+        }
+    }
+    for (log_schema_key, log_schema_index) in extra_log_schema.index.iter() {
+        if let Some(column_schema) = extra_log_schema.schema.get(*log_schema_index) {
+            schemas.insert(
+                log_schema_key,
+                (
+                    OpenTelemetryLogRecordAttrType::Log,
+                    *log_schema_index,
+                    column_schema,
+                ),
+            );
+        }
+    }
+    schemas
+}
```

Comment:
By changing the invocation to the following
```Rust
    let final_extra_schema_info = merge_schema(vec![
        (
            &extra_resource_schema,
            OpenTelemetryLogRecordAttrType::Resource,
        ),
        (&extra_scope_schema, OpenTelemetryLogRecordAttrType::Scope),
        (&extra_log_schema, OpenTelemetryLogRecordAttrType::Log),
    ]);
```
We can reduce the duplicated code.

```suggestion
fn merge_schema(
    input_schemas: Vec<(&SchemaInfo, OpenTelemetryLogRecordAttrType)>,
) -> BTreeMap<&String, (OpenTelemetryLogRecordAttrType, usize, &ColumnSchema)> {
    let mut schemas = BTreeMap::new();
    input_schemas
        .into_iter()
        .for_each(|(schema_info, attr_type)| {
            for (key, index) in schema_info.index.iter() {
                if let Some(col_schema) = schema_info.schema.get(*index) {
                    schemas.insert(key, (attr_type, *index, col_schema));
                }
            }
        });

    schemas
}
```